web/page0/src/app/runtime.service.ts

```diff
@@ -1,17 +1,191 @@
-import {Injectable} from '@angular/core';
-import {Observable, of} from 'rxjs';
-import {mockRuntimeList} from 'src/mock/runtimes';
-import {Runtime} from './runtime-interface';
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import {
+  catchError,
+  first,
+  forkJoin,
+  map,
+  mergeScan,
+  Observable,
+  of,
+  shareReplay,
+  startWith,
+  Subject,
+  tap,
+} from 'rxjs';
+import { Runtime, RuntimeAdditionalInfo } from './runtime-interface';
+import { defaultRuntimeSettings } from './settings';
+
+interface RuntimeRegisterAction {
+  type: 'register';
+  value: Runtime;
+}
+
+interface RuntimeUnregisterAction {
+  type: 'unregister';
+  value: string;
+}
+
+interface RuntimeRefreshAction {
+  type: 'refresh';
+  loading$: Subject<boolean>;
+}
+
+type RuntimeAction =
+  | RuntimeRegisterAction
+  | RuntimeUnregisterAction
+  | RuntimeRefreshAction;
 
 @Injectable({
   providedIn: 'root',
 })
 export class RuntimeService {
-  private runtimes: Observable<Runtime[]> = of(mockRuntimeList);
+  private getStoredRuntimes(): Runtime[] {
+    const runtimes = localStorage.getItem('runtimes');
+    // TODO: handle type error
+    if (runtimes) {
+      return JSON.parse(runtimes) as Runtime[];
+    }
+
+    return [];
+  }
+
+  private initialized = false;
+
+  private runtimeAction = new Subject<RuntimeAction>();
+
+  private storedRuntimes: Runtime[] = this.getStoredRuntimes();
+
+  private runtimes$: Observable<Runtime[]> = this.runtimeAction.pipe(
+    tap((action) => console.log(action)),
+    mergeScan((acc, action) => {
+      if (action.type === 'register') {
+        return of([...acc, action.value]);
+      }
+
+      if (action.type === 'unregister') {
+        return of(acc.filter((item) => item.alias !== action.value));
+      }
+
+      if (action.type === 'refresh') {
+        return forkJoin(
+          acc.map((runtime) =>
+            this.verifyRuntime(runtime.url, runtime.alias).pipe(
+              catchError((error) =>
+                of({
+                  url: runtime.url,
```

Comment:
We can't since verification failure should be handled differently:
1) During registration: the runtime should not be registered
2) During initialization/refresh: the runtime should be shown in "error" status

Reply:
Or...we can? I just changed the logic to use `status` in runtime for case 2

Reply:
Done